### PR TITLE
 Excess statistics for time series

### LINKED ISSUE #1934: Excess statistics for time series
### Feature description

For a given time series and given threshold values, find the excess statistics of the time series (total time above threshold, number of upwards crossings, average time above threshold, etc.).

### Motivation

This functionality has been developed and applied for time series analysis of intermittent fluctuations in the SOL region of fusion experiments. Some examples are listed below:
https://doi.org/10.1016/j.nme.2016.11.008
https://doi.org/10.1063/1.4947235
https://doi.org/10.1103/PhysRevE.97.012110

### Implementation strategy

Implementing this feature should be straight forward since the method is already developed and available in: https://github.com/uit-cosmo/fpp-analysis-tools/blob/main/fppanalysis/excess_statistics.py

The code still needs refactoring and appropriate tests and documentation needs to be added. I am planning to take a look at this in the upcoming days (could I be assigned to this issue?).

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gregordecristoforo**
## Description

For a given time series and given threshold values, find the excess statistics of the time series (total time above threshold, number of upwards crossings, average time above threshold, rms of times above threshold.).

For the moment the function `excess_stat` returns multiple values at once since I kept the original structure of the code as in https://github.com/uit-cosmo/fpp-analysis-tools/blob/main/fppanalysis/excess_statistics.py. I would like to hear your opinion, @namurphy,  on whether this functionality should be split into multiple functions or whether the current form is fine.

## Motivation and context

This functionality has been developed and applied for time series analysis of intermittent fluctuations in the SOL region of fusion experiments. Some examples are listed below:
https://doi.org/10.1016/j.nme.2016.11.008
https://doi.org/10.1063/1.4947235
https://doi.org/10.1103/PhysRevE.97.012110

## Related issues
Closes #1934

### CODE REVIEW

**Review by namurphy — COMMENTED** [review]
Thank you for submitting this, @gregordecristoforo!  This looks like really useful functionality, not just for plasma physics but for other fields as well.

* What would you think of building on the functionality of [`astropy.timeseries`](https://docs.astropy.org/en/stable/timeseries/index.html)?  Doing so would let us build upon functionality already in Astropy, which would also enable greater interoperability with the astronomy and heliophysics Python ecosystems.  A possible downside is that this would make it harder to 

* What would you think of having something like an `ExcessStatistics` class, with methods for each of the quantities being returned?  It could be something like...
  ```Python
  class ExcessStatistics:
      def __init__(self, times, values):  # or accept a `TimeSeries` object?
           ...
      def number_of_crossings(self):
           ...
      def hist(self, bins):
           ...
  ```
  Then we'd be able to do things like...
  ```pycon
  >>> excess_statistics = ExcessStatistics(times, values) 
  >>> excess_statistics.number_of_crossings()
  >>> excess_statistics.hist(bins=12)
  ```

Thank you again!

**Comment by gregordecristoforo** [comment]
Thanks for your feedback @namurphy !

> What would you think of having something like an `ExcessStatistics` class, with methods for each of the quantities being returned? 

That sounds like a great idea! I reformatted the code accordingly. Currently all statistics are calculate when an `ExcessStatistics` object gets instantiated sine the calculations of the statistical properties are quite entangled. What do you think about the current form?

> What would you think of building on the functionality of [astropy.timeseries](https://docs.astropy.org/en/stable/timeseries/index.html)?

Do you suggest to rewrite the code so that one can instantiate a `ExcessStatistics` object with a `TimeSeries` object as an argument? Should `ExcessStatistics` be able to handle both `astropy.timeseries` and a simple sequence (as it does now) or would you suggest writing a separate function/class for it?

**Review by namurphy — COMMENTED** [review]
> Currently all statistics are calculated when an `ExcessStatistics` object gets instantiated sine the calculations of the statistical properties are quite entangled. What do you think about the current form?

I like it! 

> Do you suggest to rewrite the code so that one can instantiate a `ExcessStatistics` object with a `TimeSeries` object as an argument? Should `ExcessStatistics` be able to handle both `astropy.timeseries` and a simple sequence (as it does now) or would you suggest writing a separate function/class for it?

I'm still thinking about how best to handle this. I'm still wondering about how best to handle that.  A possibility would be to have `ExcessStatistics` accept either a `TimeSeries` object, or a sequence that then gets converted into a `TimeSeries` object.  We'd have to be careful with uneven timesteps and might need to handle things differently when time is a `Quantity` vs when it is a NumPy array.  I'm thinking maybe we'd want to skip `TimeSeries` compatibility for this pull request, and instead perhaps raise an issue about it after this pull request gets merged.

Thank you again!

**Comment by namurphy** [comment]
Thank you for following up on this!  I think the linter issue suddenly came up because of a new release of one of our linter packages, but we just addressed that in #2046.  I think you'll need to do a `git pull` from the `main` branch.

**Comment by gregordecristoforo** [comment]
Thanks! That resolved the issue.

File: plasmapy/analysis/time_series/tests/test_excess_statistics.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+)
+def test_ExcessStatistics(signal, thresholds, time_step, pdf, bins, expected):
+    """test excess_stat function"""
+    tmp = ExcessStatistics(signal, thresholds, time_step)
+    assert tmp.total_time_above_threshold == expected[0]
+    assert tmp.number_of_crossings == expected[1]
+    assert tmp.average_times == expected[2]
+    assert tmp.rms_times == expected[3]
+    if pdf:
+        hist, bin_centers = tmp.hist(bins)
```
Comment (namurphy) [inline plasmapy/analysis/time_series/tests/test_excess_statistics.py:78]:
```suggestion
    excess_stats = ExcessStatistics(signal, thresholds, time_step)
    assert excess_stats.total_time_above_threshold == expected[0]
    assert excess_stats.number_of_crossings == expected[1]
    assert excess_stats.average_times == expected[2]
    assert excess_stats.rms_times == expected[3]
    if pdf:
        hist, bin_centers = excess_stats.hist(bins)
```
I'm wondering if a more explicit name would help improve readability.

**Review by namurphy — COMMENTED** [review]
@gregordecristoforo — apologies for the delay!  I had a few more suggestions, and then it should be ready to be merged.

Later on, we could probably expand this class to include Astropy `Quantity` objects without too much difficulty (since they operate pretty similarly to NumPy arrays) but that would be best for a different pull request if we decide to go that route.

Thank you again!

File: plasmapy/analysis/time_series/excess_statistics.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    Parameters
+    ----------
+    signal : 1D |array_like|
+        Signal to be analyzed.
+
+    thresholds : 1D |array_like|
+        Threshold values.
+
+    time_step : int
```
[the remark this answers is not in the document]
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/excess_statistics.py:25]: yes, thanks for spotting this

File: plasmapy/analysis/time_series/excess_statistics.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            The number of bins in the estimation of the PDF above ``thresholds``.
+
+        Returns
+        -------
+        hist: 2D `~numpy.ndarray`, shape (``thresholds.size``, ``bins`` )
+            For each value in ``thresholds``, returns the estimated PDF of time
+            above threshold.
+
+        bin_centers: 2D `~numpy.ndarray`, shape (``thresholds.size``, ``bins`` )
+            Bin centers for ``hist``.
```
[the remark this answers is not in the document]
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/excess_statistics.py:123]: I'll add it to the docstring

File: plasmapy/analysis/time_series/tests/test_excess_statistics.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+@pytest.mark.parametrize(
+    "signal, thresholds, time_step, bins",
+    [([1, 2], 1, -1, 32), ([1, 2], 1, 1, 1.5)],
+)
+def test_ExcessStatistics_exception(signal, thresholds, time_step, bins):
+    """test whether exception is risen"""
+    with pytest.raises((TypeError, ValueError)):
+        tmp = ExcessStatistics(signal, thresholds, time_step)
+        tmp.hist(bins)
```
[the remark this answers is not in the document]
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/tests/test_excess_statistics.py:91]: Thanks for the advice, that's much more elegant.

**Comment by gregordecristoforo** [comment]
Thanks @namurphy for your suggestions! Expanding this class to include Astropy Quantity objects sounds like an interesting idea for a future PR.

May I ask for your help one last time? Do you know the reason for the `build_docs_pins ERROR`?

**Comment by namurphy** [comment]
Yes, I am all too familiar with this error.  This is one example where Sphinx gives a not-very-helpful error message from a problem with our documentation build.

We just created a new [documentation troubleshooting guide](https://docs.plasmapy.org/en/latest/contributing/doc_guide.html). This paragraph is the most relevant.  I'll submit a pull request doing what's in the last sentence, which I thought I had done before.

> This warning sometimes occurs in the type specification of a [parameter](https://docs.python.org/3/glossary.html#term-parameter) in a docstring. Sphinx attempts to link words in type specifications to code objects. Type lines are intended to provide concise information about allowed types, sizes, shapes, physical types, and default values of a parameter. To resolve this warning, first move information about the meaning of a parameter from the type specification into the parameter description that begins on the following line. To expand the list of allowed words or patterns in type specifications, add a regular expression to `nitpick_ignore_regex` in [`docs/conf.py`](https://github.com/PlasmaPy/PlasmaPy/blob/main/docs/conf.py).

**Review by namurphy — APPROVED** [review]
LGTM!  Thank you again for doing this, @gregordecristoforo!  I'll set this to auto-merge once the tests pass.

**Pull request merged**